test/distributed/test_c10d_common.py

```diff
@@ -515,8 +518,9 @@ def test_multi_limit_multi_dtype(self):
             torch.empty([50], dtype=torch.float),
             torch.empty([25], dtype=torch.double),
         ]
-        result = dist._compute_bucket_assignment_by_size(tensors, [200, 400])
+        result, per_bucket_size_limits = dist._compute_bucket_assignment_by_size(tensors, [200, 400])
         self.assertEqual([[0], [1], [2, 4], [3, 5]], result)
+        self.assertEqual(per_bucket_size_limits, [200, 200, 400, 400])
```

Comment:
(this is prior to this PR) curious, why do we need to apply size limits independently to each dtype? IIRC, the original intention is to have a small first bucket to kick off comm earlier. If so, would it be sufficient to have just one small bucket followed by all large bucket limits? Or is this just to simply the implementation?

Reply:
> IIRC, the original intention is to have a small first bucket to kick off comm earlier

I think this is my understanding as well, but interestingly before rebuild buckets, the first bucket size actually ends up being the last bucket size because we reverse parameter ready order when passing into the reducer. So prior to rebuild buckets PR landing, we actually didn't have a smaller first bucket size according to my understanding.

@zhaojuanmao pointed out this inconsistency and potentially we should investigate it more to determine where the smaller bucket makes the most sense. 

As far as I understand, the reason we apply size limits independently to different types is so that tensors with different types go into different buckets? At least that's what I can tell from `compute_bucket_assignment_by_size`, but not sure what the motivation for that is.